src/uucore/src/lib/features/quoting_style.rs

```diff
@@ -205,144 +253,177 @@ impl Iterator for EscapedChar {
     }
 }
 
-fn shell_without_escape(name: &str, quotes: Quotes, show_control_chars: bool) -> (String, bool) {
+fn shell_without_escape(name: &[u8], quotes: Quotes, show_control_chars: bool) -> (Vec<u8>, bool) {
     let mut must_quote = false;
-    let mut escaped_str = String::with_capacity(name.len());
-
-    for c in name.chars() {
-        let escaped = {
-            let ec = EscapedChar::new_shell(c, false, quotes);
-            if show_control_chars {
-                ec
-            } else {
-                ec.hide_control()
-            }
-        };
+    let mut escaped_str = Vec::with_capacity(name.len());
+    let mut utf8_buf = vec![0; 4];
+
+    for s in name.utf8_chunks() {
+        for c in s.valid().chars() {
+            let escaped = {
+                let ec = EscapedChar::new_shell(c, false, quotes);
+                if show_control_chars {
+                    ec
+                } else {
+                    ec.hide_control()
+                }
+            };
 
-        match escaped.state {
-            EscapeState::Backslash('\'') => escaped_str.push_str("'\\''"),
-            EscapeState::ForceQuote(x) => {
-                must_quote = true;
-                escaped_str.push(x);
-            }
-            _ => {
-                for char in escaped {
-                    escaped_str.push(char);
+            match escaped.state {
+                EscapeState::Backslash('\'') => escaped_str.extend_from_slice(b"'\\''"),
+                EscapeState::ForceQuote(x) => {
+                    must_quote = true;
+                    escaped_str.extend_from_slice(x.encode_utf8(&mut utf8_buf).as_bytes());
+                }
+                _ => {
+                    for c in escaped {
+                        escaped_str.extend_from_slice(c.encode_utf8(&mut utf8_buf).as_bytes());
+                    }
                 }
             }
         }
+
+        if show_control_chars {
+            escaped_str.extend_from_slice(s.invalid());
+        } else {
+            escaped_str.resize(escaped_str.len() + s.invalid().len(), b'?');
+        }
     }
 
-    must_quote = must_quote || name.starts_with(SPECIAL_SHELL_CHARS_START);
+    must_quote = must_quote || !name.is_empty() && SPECIAL_SHELL_CHARS_START.contains(&name[0]);
```

Comment:
Maybe add a comment to explain that this does a kind of `.starts_with`.

Also, this is a personal preference, but I think adding parentheses around the `&&` is a nice to have, even if the operator precedence handles it. It's easier to understand and prevents misunderstanding.

Reply:
Good call, ended up breaking that half out into its own function instead.

Reply:
Yup, that's good for me :+1: